File: src/main/java/plus/maa/backend/service/UserSessionService.java
package plus.maa.backend.service;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.stereotype.Service;
import plus.maa.backend.repository.RedisCache;
import plus.maa.backend.service.model.LoginUser;

@Service
public class UserSessionService {

    private static final String REDIS_KEY_PREFIX_LOGIN = "LOGIN:";

    private static String buildUserCacheKey(String userId) {
        return REDIS_KEY_PREFIX_LOGIN + userId;
    }

    private final RedisCache cache;

    public UserSessionService(RedisCache cache) {
        this.cache = cache;
    }

    @Nullable
    public LoginUser getUser(String id) {
        return cache.getCache(buildUserCacheKey(id), LoginUser.class);
    }

    public void setUser(@NotNull LoginUser user) {
        cache.setCache(buildUserCacheKey(user.getUserId()), user);
    }

    public void removeUser(String id) {
        cache.removeCache(buildUserCacheKey(id));
    }

}

File: src/main/java/plus/maa/backend/service/session/UserSessionService.kt
package plus.maa.backend.service.session

import org.springframework.stereotype.Service
import plus.maa.backend.config.external.MaaCopilotProperties
import plus.maa.backend.repository.RedisCache

@Service
class UserSessionService(private val cache: RedisCache, properties: MaaCopilotProperties) {
    private val sessionExpiration = properties.jwt.refreshExpire

    fun getSession(id: String): UserSession? {
        return cache.getCache(buildUserCacheKey(id), UserSession::class.java, null, sessionExpiration)
    }

    fun setSession(id: String, session: UserSession) {
        cache.setCache(buildUserCacheKey(id), session, sessionExpiration)
    }

    fun setSession(id: String, consumer: (UserSession) -> Unit) {
        val session = UserSession()
        consumer(session)
        cache.setCache(buildUserCacheKey(id), session, sessionExpiration)
    }

    fun updateSessionIfPresent(id: String, consumer: (UserSession) -> Unit){
        cache.updateCache(id, UserSession::class.java, null, { session: UserSession? ->
            if (session != null) consumer(session)
            session
        }, sessionExpiration)
    }

    fun removeSession(id: String) {
        cache.removeCache(buildUserCacheKey(id))
    }

    companion object {
        private const val REDIS_USER_SESSION_PREFIX = "USER_SESSION_"

        private fun buildUserCacheKey(userId: String): String {
            return REDIS_USER_SESSION_PREFIX + userId
        }
    }
}
